File: fs/Practice2/DynamicProgramming/Bit/PlushToys.fs
// https://atcoder.jp/contests/joi2017yo/tasks/joi2017yo_d

stdin.ReadLine()
|> fun x -> x.Split()
|> Array.map int
|> fun arr -> (arr.[0], arr.[1])
|> fun (n, m) ->
  (
    Array.init n (fun _ -> stdin.ReadLine() |> int |> (+) -1),
    Array.create (1 <<< m) (0, 0)
  )
  |> fun (items, memo) ->
    items
    |> Seq.groupBy id
    |> Seq.fold (
      fun acc (i, v) ->
        Array.concat [ acc.[0..(i - 1)]; [| v |> Seq.length |]; acc.[(i + 1)..] ]
    ) (Array.create m 0)
    |> fun counts ->
      [1..(1 <<< m) - 1]
      |> Seq.iter (
        fun i ->
          [0..m - 1]
          |> List.filter (fun j -> i &&& (1 <<< j) = (1 <<< j))
          |> List.map (
            fun j ->
              (
                memo.[i ^^^ (1 <<< j)],
                counts.[j]
              )
              |> fun ((pc, pv), c) ->
                items.[pc..pc + c - 1]
                |> Array.filter ((<>) j)
                |> Array.length
                |> fun v -> (pc + c, pv + v)
          )
          |> List.minBy (fun (c, v) -> v)
          |> fun x -> memo.[i] <- x
      )
    |> fun _ -> memo.[(1 <<< m) - 1] |> snd
|> printfn "%d"

// ----------

// input:
// 12 4
// 1
// 3
// 2
// 4
// 2
// 1
// 2
// 3
// 1
// 1
// 3
// 4

// memo:
// [|
//   (0, 0)
//   (4, 3)
//   (3, 2)
//   (7, 4)
//   (3, 2)
//   (7, 5)
//   (6, 4)
//   (10, 6)
//   (2, 2)
//   (6, 5)
//   (5, 3)
//   (9, 5)
//   (5, 3)
//   (9, 5)
//   (8, 5)
//   (12, 7)
// |]

// output:
// 7

File: scala/Practice2/DynamicProgramming/Bit/PlushToys.scala
// https://atcoder.jp/contests/joi2017yo/tasks/joi2017yo_d

object Main {

  implicit class ScopeFunction[T](x: T) {
    def let[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    io.StdIn.readLine
      .split(" ")
      .map(_.toInt)
      .let(arr => (arr(0), arr(1)))
      .let { case (n, m) =>
        (
          Array.range(0, n).map(_x => io.StdIn.readLine.toInt.let(_ - 1)),
          Array.fill(m)(0),
          Array.range(0, m).map(_x => Array.fill(n + 1)(0)),
          Array.fill(1 << m)(0, 0)
        )
        .let { case (items, cs, vss, memo) =>
          items.foldLeft(0)((i, x) =>
            (
              (cs(x) = cs(x) + 1),
              ((0 until m).foreach(j => vss(j)(i + 1) = cs(j)))
            )
            .let(_x => i + 1)
          )
          .let(_x =>
            (1 until 1 << m).foreach(i =>
              (0 until m)
                .filter(j => (i & (1 << j)) == (1 << j))
                .map(j =>
                  (memo(i ^ (1 << j)), cs(j), vss(j))
                    .let { case ((pc, pv), c, vs) =>
                      (pc + c, pv + (c - vs(pc + c) + vs(pc)))
                    }
                )
                .minBy { case (c, v) => v }
                .let(x => memo(i) = x)
            )
          )
          .let(_x => memo((1 << m) - 1)._2)
        }
      }
      .let(println)
  )
}

// https://atcoder.jp/contests/joi2017yo/submissions/21141200
